File: C/Pikeman.c
// Does not work for Pikeman Hard because of long long overflow

#include <stdio.h>
#define MIN(x,y) (x<y)?(x):(y)

int main(){    
    long long N, T, A, B, C, i, j=0, k, q, reps, rem, ph, count=0, out=0, out2=0, curt=0, array[100005]={0}, array2 [100005]={0};
    scanf ("%lld %lld", &N, &T);
    scanf ("%lld %lld %lld %lld", &A, &B, &C, &q);
    while (array[q] == 0){
        array[q]++;
        array2[count] = q;
        q = ((A*q + B)%C)+1;
        count++;
    }

    while (array2[j] != q){ //j is index where it loops
        j++;
    }
    k = count-j; //k is the length of the loop

    reps = (N-j)/k;
    rem = (N-j)%k;
    for (i=j; i<count; i++){
        array[array2[i]] = reps;
        if (i-j < rem){
            array[array2[i]]++;
        }
    }

    i=1;
    while (T > i){
        if (out == N){
            break;
        }
        ph = MIN(array[i], T/i);
        out += ph;
        T -= ph*i;
        out2 = (out2 + curt*ph+((ph*(ph+1)*i)/2))%1000000007;
        curt += ph*i;
        i++;
    }
    printf("%lld %lld\n", out, out2);
    return 0;
}

File: Python/Pikeman.py
qq = raw_input()
rr = raw_input()
dlist = [int (x) for x in qq.split()]
dlist2 = [int(x) for x in rr.split()]
array = [0 for x in range (100005)]
array2 = [0 for x in range (100005)]

q = dlist2[3]
count = 0
j = 0
out = 0
out2 = 0
curt = 0

while array[q] == 0:
    array[q] = 1
    array2[count] = q
    q = (dlist2[0]*q + dlist2[1])%dlist2[2]+1
    count += 1

while (array2[j] != q):
    j += 1
k = count - j

reps = (dlist[0]-j)/k
rem = (dlist[0]-j)%k

for i in range(j, count):
    array[array2[i]] = reps
    if (i-j < rem):
        array[array2[i]] += 1

i=1
while (dlist[1] > i):
    if (out == dlist[0]):
        break
    ph = min(array[i], dlist[1]/i)
    out += ph
    dlist[1] -= ph*i
    out2 = (out2 + curt*ph + (ph*(ph+1)*i)/2)%1000000007
    curt += ph*i
    i += 1

print str(out), str(out2)

